### PR TITLE
turn on -Werror=unused-variable in our Bazel CPU build

### PR DESCRIPTION
**Pull request by dagitses**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #79327
* #79306
* #79305
* #79313
* #79312
* #79308
* __->__ #79156

Summary:
We also fix any existing issues. Note that we only do this for the CPU
build because nvcc is considered a C++ toolchain but it does not have
the same flag support. Adding flags to the GPU build will cause nvcc
errors.

Test Plan: Built locally, rely on CI to confirm.

Reviewers: malfet

Subscribers:

Tasks:

Tags:

### CODE REVIEW

**Review by seemethere — APPROVED** [review]

**Review by osalpekar — APPROVED** [review]

File: c10/test/core/DispatchKeySet_test.cpp [hunk 1]
```diff
@@ -370,29 +367,12 @@ TEST(DispatchKeySet, IteratorCrossProduct) {
...
-}
-
-TEST(DispatchKeySet, IteratorRangeFull) {
-  DispatchKeySet full_set(DispatchKeySet::FULL);
-  uint8_t i = 0;
-
-  for (DispatchKey dispatch_key : full_set) {
-    i++;
-  }
+  std::ptrdiff_t count = std::distance(full_set.begin(), full_set.end());
```
Comment (albanD) [inline c10/test/core/DispatchKeySet_test.cpp:370]: Is this doing the same thing?
The goal here is really to go through the whole iterator.
Also the goal was to explicitely test that both iteration types are working.

cc @bdhirsh

File: c10/test/util/intrusive_ptr_test.cpp [hunk 2]
```diff
@@ -218,8 +218,6 @@ TEST(
     givenInvalidPtr_whenMoveAssigning_thenNewInstanceIsValid) {
   intrusive_ptr<SomeClass> obj1 = make_intrusive<SomeClass>();
   intrusive_ptr<SomeClass> obj2;
-  // NOLINTNEXTLINE(clang-analyzer-deadcode.DeadStores)
```
Comment (albanD) [inline c10/test/util/intrusive_ptr_test.cpp:221]: I'm not sure the changes here are ok. The whole point of these tests is to observe the side effect of this code I think.

cc @ezyang

File: c10/test/core/DispatchKeySet_test.cpp (hunk 1 above)
Comment (dagitses) [reply c10/test/core/DispatchKeySet_test.cpp:370]: I think it must. It's not a random access iterator so distance must advance one at a time until it reaches the end.

File: c10/test/util/intrusive_ptr_test.cpp (hunk 2 above)
Comment (dagitses) [reply c10/test/util/intrusive_ptr_test.cpp:221]: I don't think `get()` has side effects? It just returns the pointer.

Comment (albanD) [reply c10/test/util/intrusive_ptr_test.cpp:221]: But the `.weak` below do right?

Comment (dagitses) [reply c10/test/util/intrusive_ptr_test.cpp:221]: Oh you're right. Thanks, let me add those and the locks back in.

Comment (dagitses) [reply c10/test/util/intrusive_ptr_test.cpp:221]: Thanks, back in.

**Review by albanD — APPROVED** [review]
SGTM

**Comment by albanD** [comment]
@dagitses wrong bot ;)

**Comment by dagitses** [comment]
> @dagitses wrong bot ;)

gosh darn it. i'm one of the last hold outs on the internal first workflow.

**Pull request closed without merging**